### PR TITLE
fix NaN buffer samples caused by fractional index when filter_l is odd

### PR DESCRIPTION
**Pull request by rkramer1964**
Using chrome/firefox under windows with a filter_l of 31 caused an index (j2) to be computed as nn.5 (e.g. -29.5) resulting in and undefined y and NaN for xvalue in fill_buffer_resample.  This caused silence to be returned when encoding mp3 samples from PCM.

Noted in Rocket.Chat when trying to record audio messages under windows.

### CODE REVIEW

File: src/js/Lame.js [hunk 1]
```diff
@@ -1786,7 +1786,8 @@ function Lame() {
             joff = 0 | Math.floor((offset * 2 * bpc) + bpc + .5);
             var xvalue = 0.;
             for (i = 0; i <= filter_l; ++i) {
-                var j2 = i + j - filter_l / 2;
+		/* force integer index */
+                var j2 = Math.floor(i + j - filter_l / 2);
```
Comment (zhuker) [inline src/js/Lame.js]: i recommend `var j2 = 0 | (i + j - filter_l / 2);` instead - it's faster

**Comment by rkramer1964** [comment]
I disagree.  If you 1) look at line 1794, you allow for negative  
indices.  0 would be incorrect in that case; 2) the point of the patch  
is to remove the fraction from the index.  Consider order of  
operations when filter_l is 31 (as it is in chrome v 66+ under  
windows).  It works out to i + j - 15.5.  Since i and j are integer,  
you get a non-integer value for j2 which is then applied as the index  
on 1794.

Disallow the pull if you like, but the code you recommend still leaves  
j2 fractional and results in an incorrect xvalue at line 1796.  The  
better option would be for force filter_l to always be an even number,  
but you test for oddness at the beginning of the function which I am  
trying to leave undisturbed.

Rick

Quoting Alex Zhukov <notifications@github.com>:

> @ZHUKER requested changes on this pull request.
>
> -------------------------
>
>   In src/js/Lame.js[1]:
>> @@ -1786,7 +1786,8 @@ function Lame() {              joff = 0 |  
>> Math.floor((offset * 2 * bpc) + bpc + .5);              var xvalue  
>> = 0.;              for (i = 0; i <= filter_l; ++i) { -               
>>   var j2 = i + j - filter_l / 2; +  /* force integer index */ +      
>>            var j2 = Math.floor(i + j - filter_l / 2);
>   i recommend var j2 = 0 | (i + j - filter_l / 2); instead - it's faster
>
>   —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub[2], or mute the thread[3].



Links:
------
[1] https://github.com/zhuker/lamejs/pull/57#discussion_r256091856
[2] https://github.com/zhuker/lamejs/pull/57#pullrequestreview-202839438
[3]

**Comment by rkramer1964** [comment]
I was wrong.  I misread the | as ||.  Too many hours in front of the screen.

My apologies.

Quoting Alex Zhukov <notifications@github.com>:

> @ZHUKER requested changes on this pull request.
>
> -------------------------
>
>   In src/js/Lame.js[1]:
>> @@ -1786,7 +1786,8 @@ function Lame() {              joff = 0 |  
>> Math.floor((offset * 2 * bpc) + bpc + .5);              var xvalue  
>> = 0.;              for (i = 0; i <= filter_l; ++i) { -               
>>   var j2 = i + j - filter_l / 2; +  /* force integer index */ +      
>>            var j2 = Math.floor(i + j - filter_l / 2);
>   i recommend var j2 = 0 | (i + j - filter_l / 2); instead - it's faster
>
>   —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub[2], or mute the thread[3].



Links:
------
[1] https://github.com/zhuker/lamejs/pull/57#discussion_r256091856
[2] https://github.com/zhuker/lamejs/pull/57#pullrequestreview-202839438
[3]

File: src/js/Lame.js (hunk 1 above)
Comment (rkramer1964) [reply src/js/Lame.js]: I'll get it changed momentarily

**Comment by rkramer1964** [comment]
accidentally closed

**Pull request merged**